src/dataset/mod.rs

```diff
@@ -0,0 +1,62 @@
+use lightgbm_sys::DatasetHandle;
+#[cfg(feature = "dataframe")]
+use polars::prelude::DataFrame;
+
+use OutputVec;
+use {Error, InputMatrix};
+
+#[cfg(feature = "dataframe")]
+mod dataframe;
+mod ffi;
+
+/// Represents an unloaded Dataset for the Booster Builder.
+/// At the fit step of the BoosterBuilder, these will be added to the
+/// lightgbm backend
+#[derive(Clone)]
+pub struct DataSet {
+    format: DataFormat,
+    /// Possible params for the Dataset. This is !currently! not in use
+    /// and will be just an empty string
+    params: String,
+}
+
+#[derive(Clone)]
+pub enum DataFormat {
+    File {
+        path: String,
+    },
+    Vecs {
+        x: InputMatrix,
+        y: OutputVec,
+    },
+    #[cfg(feature = "dataframe")]
+    DataFrame {
+        df: DataFrame,
+        y_column: String,
+    },
+}
```

Comment:
I guess this is to make Datasets clonable?

It feels like it makes datasets and error handling a bit more complicated, compared to just directly loading them. It would also prevent future load_* functions that only take a reference to properly laid-out data (maybe loading nalgebra arrays,if the support that?).

I think (suspect) you can implement clone on Dataset by calling `LGBM_DatasetCreateByReference(h_old, rows, &mut h_new)` followed by `LGBM_DatasetAddFeaturesFrom(h_new, h_old)`. But maybe that's completely wrong, the documentation is incredibly vague.